### PR TITLE
50 bug wrong response type for glom

### PR DESCRIPTION
**Pull request by RMeissnerCC**
Due to an unexpected entry in general_keywords for collection "Informatik", this endpoint failed in production and had to be mended.
The corresponding endpoint was transfered here.I refactored specificially enum unions which caused their entries to be hidden, the naming of certain elasticsearch code and generalized the usage of glom - the later could be replaced by our own implementation if we want to reduce dependencies.

### CODE REVIEW

File: src/app/api/collections/missing_attributes.py [hunk 1]
```diff
@@ -0,0 +1,87 @@
...
+    search = missing_attributes_search(noderef_id, missing_attribute, max_hits)
+
+    response = search.execute()
+    if response.success():
+        return map_elastic_response_to_model(
+            response, missing_attributes_spec, MissingMaterials
+        )
+
+
+missingPropertyFilter = [
```
Comment (MRuecklCC) [inline src/app/api/collections/missing_attributes.py:82]: move to top or the file where it is used?

File: src/app/api/collections/models.py [hunk 2]
```diff
@@ -13,3 +13,11 @@ class CollectionNode(BaseModel):
     title: Optional[str]  # might be none due to data model
     children: list[CollectionNode]
     parent_id: Optional[UUID]
+
+
+class MissingMaterials(CollectionNode):
```
Comment (MRuecklCC) [inline src/app/api/collections/models.py:18]: some documentation what the actual attributes are would be nice :P

File: src/app/api/collections/utils.py [hunk 3]
```diff
@@ -0,0 +1,19 @@
...
+from elasticsearch_dsl.response import Response
+from glom import glom
+
+from app.api.collections.models import CollectionNode, MissingMaterials
+
+T = TypeVar("T", CollectionNode, MissingMaterials)
+
+
+def map_elastic_response_to_model(
+    response: Response, specs: dict, model: Generic[T]
```
Comment (MRuecklCC) [inline src/app/api/collections/utils.py:12]: not sure if `Generic[T]` is the right type annotation here. eventually it is `Type[T]`. Does pycharm complain when the function is called in `tree.py` ?

File: src/app/api/score/models.py [hunk 4]
```diff
@@ -2,35 +2,35 @@
 
 from pydantic import BaseModel, Field
 
-from app.elastic.fields import Field as ElasticField
-from app.elastic.fields import FieldType
+from app.elastic.fields import ElasticField as ElasticField
```
Comment (MRuecklCC) [inline src/app/api/score/models.py:5]: redundant as

File: src/app/api/collections/utils.py [hunk 5]
```diff
@@ -0,0 +1,19 @@
...
+) -> list[T]:
+    collections = []
+    for hit in response:
+        parsed_entry = glom(hit.to_dict(), specs)
+        if parsed_entry["title"] is not None:
```
Comment (MRuecklCC) [inline src/app/api/collections/utils.py:17]: why would there be entries with title = None? Would that be a bug in the elastic query? Eventually add a comment why those scenarios do not matter.

File: src/app/models.py [hunk 6]
```diff
@@ -1,40 +1,30 @@
...
-
-
-CollectionAttribute = Field(
-    "CollectionAttribute",
-    [
-        (f.name, (f.value, f.field_type))
-        for f in chain(ElasticResourceAttribute, _CollectionAttribute)
-    ],
-)
+class ElasticResourceAttribute(ElasticField):
```
Comment (MRuecklCC) [inline src/app/models.py:11]: What is the conceptual difference between the `ElasticResourceAttribute`s and the `CollectionAttribute`s? 

I would have expected one of them to only specify attributes from e.g.` properties`, whereas the other one only takes attributes from `nodeRef`.

File: tests/unit_tests/crud/test_missing_attributes.py [hunk 7]
```diff
@@ -0,0 +1,96 @@
+import uuid
+
+from elasticsearch_dsl import AttrDict
+
+from app.api.collections.missing_attributes import (
+    missing_attributes_search,
+    missing_attributes_spec,
+    missingPropertyFilter,
```
Comment (MRuecklCC) [inline tests/unit_tests/crud/test_missing_attributes.py:8]: if this is only used here, then we can as well inline it here! :)

File: tests/unit_tests/crud/test_missing_attributes.py [hunk 8]
```diff
@@ -0,0 +1,96 @@
...
+def test_missing_attributes_search():
+    dummy_uuid = uuid.uuid4()
+    dummy_attribute = "properties.cm:title"
+    dummy_missing_attribute = missingPropertyFilter[0].value
+    print(dummy_missing_attribute)
+    dummy_maximum_size = 3
+    search = missing_attributes_search(
+        dummy_uuid, dummy_missing_attribute, dummy_maximum_size
+    )
+    assert search.to_dict() == {
```
Comment (MRuecklCC) [inline tests/unit_tests/crud/test_missing_attributes.py:23]: Eventually it makes sense to just kick out the whole elastic dsl depenency...

I mean, there is tons framework / utility code to build a hand full of different queries, and then there is tests that test that framework code, where we could just build the queries exactly like demonstrated in that test here.

That will make stuff way more consolidated and maintainable, eventually on the cost of a little repetition... It's a tradeof :thinking:

**Review by MRuecklCC — CHANGES_REQUESTED** [review]
Some minor things.

The biggest point (which isn't realls in the scope of this PR) is the following:

Currently there is a huge machinery around building elastic search request and processing the responses.

We have a separation into two different enums (ElasticResourceAttribute, and CollectionAttribute) which seems entirely arbitrary to me. It also seems that the separation into enums has to be aligned with some other internal data structures (the request building and the internal data structures into which the responses will be transformed).

As far as I understood the overall procedure is about the following

- Issue request to elastic yielding a deeply nested request specific json
- use the pair of (glom-specs, `...Attribute` enum) to transform the response dictionary into the request specific internal representation which would be `list[MissingMaterials]` or `list[CollectionNode]`

Currently this whole process is scattered over a quite large amount of different files. It would be nice to have that all together, i.e. request building, response specs, and internal data model definition.

But as stated above, thats definitely out of scope for this PR...

File: src/app/api/api.py [hunk 9]
```diff
@@ -57,6 +61,7 @@ def get_database(request: Request) -> Database:
     """
 
 TAG_STATISTICS = "Statistics"
```
[the remark this answers is not in the document]
Comment (RMeissnerCC) [reply src/app/api/api.py:63]: Yes, for all endpoints related to quality. I could rename it to "Quality" or similar.

File: src/app/api/collections/missing_attributes.py (hunk 1 above)
Comment (RMeissnerCC) [reply src/app/api/collections/missing_attributes.py:82]: Aaaaaahhh, snake case it must be. I moved it to the top.

File: src/app/api/collections/models.py (hunk 2 above)
Comment (RMeissnerCC) [reply src/app/api/collections/models.py:18]: Added docstring and comments

File: src/app/api/collections/utils.py (hunk 3 above)
Comment (RMeissnerCC) [reply src/app/api/collections/utils.py:12]: No, it does not. But it does complain with Type[T]. I generally check that IDE, tests and using the actual endpoints does not raise an exception and works as I believe it should.

File: src/app/api/score/models.py (hunk 4 above)
Comment (RMeissnerCC) [reply src/app/api/score/models.py:5]: Removed

File: src/app/models.py [hunk 10]
```diff
@@ -1,40 +1,30 @@
...
+    EDU_CONTEXT = ("properties.ccm:educationalcontext", ElasticFieldType.TEXT)
```
[the remark this answers is not in the document]
Comment (RMeissnerCC) [reply src/app/models.py:12]: Yes :D I got confused over and over :D

File: src/app/api/collections/utils.py (hunk 5 above)
Comment (RMeissnerCC) [reply src/app/api/collections/utils.py:17]: Ayayay, outdated logic, I cleaned it :)

File: src/app/models.py (hunk 6 above)
Comment (RMeissnerCC) [reply src/app/models.py:11]: This is historic and will be part of future PR's when I continue to merge . I added a TODO here.

File: tests/unit_tests/crud/test_missing_attributes.py (hunk 7 above)
Comment (RMeissnerCC) [reply tests/unit_tests/crud/test_missing_attributes.py:8]: It is only used in the API, but for two different cases, see the new endpoint. I am unsure of a good way to inline it there and I would like to keep as much details out of the api.py as possible.
Oh and it is used in the tests as well :D

File: tests/unit_tests/crud/test_missing_attributes.py (hunk 8 above)
Comment (RMeissnerCC) [reply tests/unit_tests/crud/test_missing_attributes.py:23]: I agree partially, a lot of unnecessary code is built around elastic - but mostly by the former maintainer. If we want to get rid of elasticsearch_dsl, we would have to drop the complete Search() etc. stuff, i.e., implement our own algorithm to talk with Elasticsearch. Currently, I see no need to do just that. But reducing complexity is high on my list here.

**Comment by RMeissnerCC** [comment]
> Some minor things.
> 
> The biggest point (which isn't realls in the scope of this PR) is the following:
> 
> Currently there is a huge machinery around building elastic search request and processing the responses.
> 
> We have a separation into two different enums (ElasticResourceAttribute, and CollectionAttribute) which seems entirely arbitrary to me. It also seems that the separation into enums has to be aligned with some other internal data structures (the request building and the internal data structures into which the responses will be transformed).
> 
> As far as I understood the overall procedure is about the following
> 
> * Issue request to elastic yielding a deeply nested request specific json
> * use the pair of (glom-specs, `...Attribute` enum) to transform the response dictionary into the request specific internal representation which would be `list[MissingMaterials]` or `list[CollectionNode]`
> 
> Currently this whole process is scattered over a quite large amount of different files. It would be nice to have that all together, i.e. request building, response specs, and internal data model definition.
> 
> But as stated above, thats definitely out of scope for this PR...

The spreading of code over many files is a valuable observation - I had not bothered yet. This will be part of the further refactoring of elasticsearch for the next endpoints. But removing it completely, i.e., elasticsearch-dsl, I am not a fan of right now. I suggest we have a chat on this when I continue working here, alright?

File: src/app/api/api.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (MRuecklCC) [reply src/app/api/api.py:63]: just wanted to make sure we don't have unused constants (those are the worst :P)

File: src/app/api/collections/utils.py (hunk 3 above)
Comment (MRuecklCC) [reply src/app/api/collections/utils.py:12]: oki :)

File: src/app/models.py (hunk 6 above)
Comment (MRuecklCC) [reply src/app/models.py:13]: :+1:

**Pull request merged**